File: frontend/src/lib/paperEngine.test.ts
import { describe, expect, it } from "vitest";

import {
  newAccount,
  placeOrder,
  onTick,
  cancelOrder,
  resetAccount,
  unrealized,
  equity,
  positionMtm,
  DEFAULT_STARTING_CASH,
  type PaperAccount,
} from "./paperEngine";

const SYM = 1;

describe("newAccount", () => {
  it("seeds cash + startingCash and is empty", () => {
    const a = newAccount(100_000);
    expect(a.cash).toBe(100_000);
    expect(a.startingCash).toBe(100_000);
    expect(a.realizedPnl).toBe(0);
    expect(a.orders).toHaveLength(0);
    expect(a.fills).toHaveLength(0);
    expect(Object.keys(a.positions)).toHaveLength(0);
  });
  it("defaults + guards bad input", () => {
    expect(newAccount().cash).toBe(DEFAULT_STARTING_CASH);
    expect(newAccount(-5).cash).toBe(DEFAULT_STARTING_CASH);
    expect(newAccount(0).cash).toBe(DEFAULT_STARTING_CASH);
    expect(newAccount(NaN).cash).toBe(DEFAULT_STARTING_CASH);
  });
});

describe("market orders", () => {
  it("buy market fills immediately, moves cash + opens long", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order, fill } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "market", qty: 10 },
      100,
    );
    expect(order.status).toBe("filled");
    expect(fill).toBeDefined();
    expect(fill!.price).toBe(100);
    expect(fill!.qty).toBe(10);
    expect(a1.cash).toBe(100_000 - 100 * 10);
    expect(a1.positions[SYM]).toEqual({ qty: 10, avgEntry: 100 });
    expect(a1.realizedPnl).toBe(0);
    // Immutability: original untouched.
    expect(a0.cash).toBe(100_000);
    expect(a0.positions[SYM]).toBeUndefined();
  });

  it("sell market with no position opens a short and adds cash", () => {
    const a0 = newAccount(100_000);
    const { account: a1 } = placeOrder(
      a0,
      { symbolId: SYM, side: "sell", type: "market", qty: 5 },
      200,
    );
    expect(a1.cash).toBe(100_000 + 200 * 5);
    expect(a1.positions[SYM]).toEqual({ qty: -5, avgEntry: 200 });
  });

  it("market order with no market price is cancelled, no state change", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "market", qty: 5 },
      undefined,
    );
    expect(order.status).toBe("cancelled");
    expect(a1.cash).toBe(100_000);
    expect(a1.positions[SYM]).toBeUndefined();
  });

  it("non-positive qty is rejected", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "market", qty: 0 },
      100,
    );
    expect(order.status).toBe("cancelled");
    expect(a1.cash).toBe(100_000);
  });
});

describe("limit orders", () => {
  it("buy limit BELOW market rests as working", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order, fill } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "limit", price: 90, qty: 3 },
      100,
    );
    expect(order.status).toBe("working");
    expect(fill).toBeUndefined();
    expect(a1.cash).toBe(100_000);
    expect(a1.positions[SYM]).toBeUndefined();
  });

  it("buy limit AT/ABOVE market crosses and fills at the limit price", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order, fill } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "limit", price: 110, qty: 2 },
      100,
    );
    expect(order.status).toBe("filled");
    expect(fill!.price).toBe(110);
    expect(a1.cash).toBe(100_000 - 110 * 2);
    expect(a1.positions[SYM]).toEqual({ qty: 2, avgEntry: 110 });
  });

  it("sell limit ABOVE market rests as working", () => {
    const a0 = newAccount(100_000);
    const { order } = placeOrder(
      a0,
      { symbolId: SYM, side: "sell", type: "limit", price: 120, qty: 2 },
      100,
    );
    expect(order.status).toBe("working");
  });

  it("sell limit AT/BELOW market crosses and fills", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order, fill } = placeOrder(
      a0,
      { symbolId: SYM, side: "sell", type: "limit", price: 90, qty: 2 },
      100,
    );
    expect(order.status).toBe("filled");
    expect(fill!.price).toBe(90);
    expect(a1.positions[SYM]).toEqual({ qty: -2, avgEntry: 90 });
  });
});

describe("onTick — fills working orders as the market moves", () => {
  it("buy limit does NOT fill until price drops to/below the limit", () => {
    const a0 = newAccount(100_000);
    const { account: a1 } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "limit", price: 90, qty: 4 },
      100,
    );
    // Still above -> no fill.
    const a2 = onTick(a1, SYM, 95);
    expect(a2.positions[SYM]).toBeUndefined();
    expect(a2.orders[0]?.status).toBe("working");
    // Price crosses down -> fills at the limit (90), not the tick (89).
    const a3 = onTick(a2, SYM, 89);
    expect(a3.positions[SYM]).toEqual({ qty: 4, avgEntry: 90 });
    expect(a3.cash).toBe(100_000 - 90 * 4);
    expect(a3.orders[0]?.status).toBe("filled");
  });

  it("sell limit fills when price rises to/above the limit", () => {
    const a0 = newAccount(100_000);
    const { account: a1 } = placeOrder(
      a0,
      { symbolId: SYM, side: "sell", type: "limit", price: 120, qty: 3 },
      100,
    );
    const a2 = onTick(a1, SYM, 119);
    expect(a2.positions[SYM]).toBeUndefined();
    const a3 = onTick(a2, SYM, 125);
    expect(a3.positions[SYM]).toEqual({ qty: -3, avgEntry: 120 });
    expect(a3.orders[0]?.status).toBe("filled");
  });

  it("only ticks the matching symbol and ignores non-working orders", () => {
    const a0 = newAccount(100_000);
    const { account: a1 } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "limit", price: 90, qty: 4 },
      100,
    );
    const untouched = onTick(a1, 999, 50);
    expect(untouched).toBe(a1); // no working order for that symbol -> same ref
  });
});

describe("average-cost + realized PnL", () => {
  it("weighted-averages entry when adding to a long", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: SYM, side: "buy", type: "market", qty: 10 }, 100).account;
    a = placeOrder(a, { symbolId: SYM, side: "buy", type: "market", qty: 10 }, 200).account;
    // (100*10 + 200*10)/20 = 150
    expect(a.positions[SYM]).toEqual({ qty: 20, avgEntry: 150 });
  });

  it("realizes PnL on a partial close, keeps avgEntry", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: SYM, side: "buy", type: "market", qty: 10 }, 100).account;
    a = placeOrder(a, { symbolId: SYM, side: "sell", type: "market", qty: 4 }, 150).account;
    // realized = (150-100)*4 = 200
    expect(a.realizedPnl).toBe(200);
    expect(a.positions[SYM]).toEqual({ qty: 6, avgEntry: 100 });
    const lastFill = a.fills[a.fills.length - 1];
    expect(lastFill?.realized).toBe(200);
  });

  it("realizes on a full close and goes flat", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: SYM, side: "buy", type: "market", qty: 10 }, 100).account;
    a = placeOrder(a, { symbolId: SYM, side: "sell", type: "market", qty: 10 }, 130).account;
    expect(a.realizedPnl).toBe((130 - 100) * 10);
    expect(a.positions[SYM]).toBeUndefined();
  });

  it("flips long -> short cleanly, realizing only the closed leg", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: SYM, side: "buy", type: "market", qty: 10 }, 100).account;
    // Sell 15 at 120: close 10 (realize (120-100)*10=200) + open short 5 @ 120.
    a = placeOrder(a, { symbolId: SYM, side: "sell", type: "market", qty: 15 }, 120).account;
    expect(a.realizedPnl).toBe(200);
    expect(a.positions[SYM]).toEqual({ qty: -5, avgEntry: 120 });
  });

  it("realizes profit on covering a short", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: SYM, side: "sell", type: "market", qty: 10 }, 200).account;
    // Buy back 10 at 150: short profit (200-150)*10 = 500.
    a = placeOrder(a, { symbolId: SYM, side: "buy", type: "market", qty: 10 }, 150).account;
    expect(a.realizedPnl).toBe(500);
    expect(a.positions[SYM]).toBeUndefined();
  });
});

describe("unrealized / equity / positionMtm", () => {
  it("unrealized reflects mark vs avg entry (long & short)", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: 1, side: "buy", type: "market", qty: 10 }, 100).account;
    a = placeOrder(a, { symbolId: 2, side: "sell", type: "market", qty: 5 }, 200).account;
    // long: (150-100)*10 = 500 ; short: (180-200)*-5 = 100
    const marks = { 1: 150, 2: 180 };
    expect(unrealized(a, marks)).toBe(600);
  });

  it("equity = cash + Σ position MTM", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: 1, side: "buy", type: "market", qty: 10 }, 100).account;
    // cash = 1,000,000 - 1000 = 999,000 ; MTM at 150 = 1500 ; equity = 1,000,500
    expect(equity(a, { 1: 150 })).toBe(1_000_500);
  });

  it("positions without a mark contribute 0", () => {
    let a: PaperAccount = newAccount(1_000_000);
    a = placeOrder(a, { symbolId: 1, side: "buy", type: "market", qty: 10 }, 100).account;
    expect(unrealized(a, {})).toBe(0);
    expect(equity(a, {})).toBe(a.cash);
  });

  it("positionMtm handles flat / missing mark", () => {
    expect(positionMtm({ qty: 0, avgEntry: 0 }, 100)).toBe(0);
    expect(positionMtm({ qty: 5, avgEntry: 10 }, undefined)).toBe(0);
    expect(positionMtm({ qty: 5, avgEntry: 10 }, 20)).toBe(100);
  });
});

describe("cancelOrder", () => {
  it("cancels a working order and stops it filling on tick", () => {
    const a0 = newAccount(100_000);
    const { account: a1, order } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "limit", price: 90, qty: 4 },
      100,
    );
    const a2 = cancelOrder(a1, order.id);
    expect(a2.orders[0]?.status).toBe("cancelled");
    const a3 = onTick(a2, SYM, 80);
    expect(a3.positions[SYM]).toBeUndefined();
  });

  it("no-op for unknown / already-terminal orders", () => {
    const a0 = newAccount(100_000);
    expect(cancelOrder(a0, "nope")).toBe(a0);
    const { account: a1, order } = placeOrder(
      a0,
      { symbolId: SYM, side: "buy", type: "market", qty: 1 },
      100,
    );
    expect(cancelOrder(a1, order.id)).toBe(a1); // filled -> unchanged
  });
});

describe("resetAccount", () => {
  it("returns a fresh empty account with the given starting cash", () => {
    const a = resetAccount(50_000);
    expect(a.cash).toBe(50_000);
    expect(a.startingCash).toBe(50_000);
    expect(a.orders).toHaveLength(0);
    expect(a.fills).toHaveLength(0);
  });
  it("defaults when no cash passed", () => {
    expect(resetAccount().cash).toBe(DEFAULT_STARTING_CASH);
  });
});

File: android/app/src/test/java/com/testlogon/android/feature/paper/PaperEngineTest.kt
package com.testlogon.android.feature.paper

import com.testlogon.android.data.exchange.OrderSide
import com.testlogon.android.feature.paper.PaperEngine.PaperOrder
import com.testlogon.android.feature.paper.PaperEngine.PaperOrderStatus
import com.testlogon.android.feature.paper.PaperEngine.PaperOrderType
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Thorough unit tests for the PURE [PaperEngine]: market fills (buy & sell), limit fills (working ->
 * triggered, buy & sell), average-cost realized PnL, weighted-average entry, short-selling, position
 * flip, cash legs, equity/unrealized marks, and cancel.
 */
class PaperEngineTest {

    private val BTC = 1
    private val ETH = 2

    private fun order(
        id: String,
        side: OrderSide,
        type: PaperOrderType,
        qty: Long,
        limitPrice: Long? = null,
        symbolId: Int = BTC,
    ) = PaperOrder(id = id, symbolId = symbolId, side = side, type = type, qty = qty, limitPrice = limitPrice)

    // ---- market fills ----

    @Test
    fun marketBuy_fillsImmediately_debitsCash_opensLong() {
        val acct = PaperEngine.newAccount(1_000_000)
        val after = PaperEngine.placeOrder(acct, order("o1", OrderSide.BUY, PaperOrderType.MARKET, 3), 100)
        val pos = after.positions.getValue(BTC)
        assertEquals(3L, pos.qty)
        assertEquals(100L, pos.avgEntry)
        assertEquals(1_000_000 - 300, after.cash)
        assertEquals(1, after.fills.size)
        assertEquals(PaperOrderStatus.FILLED, after.orders.single().status)
    }

    @Test
    fun marketSell_fromFlat_opensShort_creditsCash() {
        val acct = PaperEngine.newAccount(1_000_000)
        val after = PaperEngine.placeOrder(acct, order("s1", OrderSide.SELL, PaperOrderType.MARKET, 2), 100)
        val pos = after.positions.getValue(BTC)
        assertEquals(-2L, pos.qty)
        assertEquals(100L, pos.avgEntry)
        assertEquals(1_000_000 + 200, after.cash)
    }

    // ---- average-cost realized (round trip) ----

    @Test
    fun buyThenSellHigher_realizesProfit_flatClosesPosition() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 10), 100)
        acct = PaperEngine.placeOrder(acct, order("s", OrderSide.SELL, PaperOrderType.MARKET, 10), 130)
        // realized = (130-100)*10 = 300
        assertEquals(300L, acct.realizedPnl)
        assertNull(acct.positions[BTC])
        // cash: -1000 (buy) + 1300 (sell) = +300 over start
        assertEquals(1_000_000 + 300, acct.cash)
    }

    @Test
    fun weightedAverageEntry_acrossTwoBuys() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b1", OrderSide.BUY, PaperOrderType.MARKET, 10), 100)
        acct = PaperEngine.placeOrder(acct, order("b2", OrderSide.BUY, PaperOrderType.MARKET, 30), 200)
        // avg = (10*100 + 30*200) / 40 = 7000/40 = 175
        assertEquals(175L, acct.positions.getValue(BTC).avgEntry)
        assertEquals(40L, acct.positions.getValue(BTC).qty)
    }

    @Test
    fun partialClose_realizesOnClosedQtyOnly_keepsRemainingAtSameAvg() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 10), 100)
        acct = PaperEngine.placeOrder(acct, order("s", OrderSide.SELL, PaperOrderType.MARKET, 4), 150)
        // realized on 4: (150-100)*4 = 200; remaining 6 @ avg 100
        assertEquals(200L, acct.realizedPnl)
        assertEquals(6L, acct.positions.getValue(BTC).qty)
        assertEquals(100L, acct.positions.getValue(BTC).avgEntry)
    }

    // ---- short realized ----

    @Test
    fun shortThenCoverLower_realizesProfit() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("s", OrderSide.SELL, PaperOrderType.MARKET, 5), 200)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 5), 120)
        // short realized = (120-200)*5*(-1) = 400
        assertEquals(400L, acct.realizedPnl)
        assertNull(acct.positions[BTC])
    }

    // ---- flip ----

    @Test
    fun sellThroughLong_flipsToShort_reseedsAvgAtFill() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 10), 100)
        acct = PaperEngine.placeOrder(acct, order("s", OrderSide.SELL, PaperOrderType.MARKET, 15), 130)
        // close 10 long @ (130-100)*10 = 300 realized; residual 5 opens short @ 130
        assertEquals(300L, acct.realizedPnl)
        val pos = acct.positions.getValue(BTC)
        assertEquals(-5L, pos.qty)
        assertEquals(130L, pos.avgEntry)
    }

    // ---- limit fills via onTick ----

    @Test
    fun limitBuy_belowMarket_rests_thenFillsWhenPriceDrops() {
        var acct = PaperEngine.newAccount(1_000_000)
        // market is 100, buy limit 90 is not marketable -> works
        acct = PaperEngine.placeOrder(acct, order("lb", OrderSide.BUY, PaperOrderType.LIMIT, 5, 90), 100)
        assertTrue(acct.positions.isEmpty())
        assertEquals(PaperOrderStatus.WORKING, acct.orders.single().status)
        // price ticks down to 90 -> fills at limit 90
        acct = PaperEngine.onTick(acct, BTC, 90)
        assertEquals(5L, acct.positions.getValue(BTC).qty)
        assertEquals(90L, acct.positions.getValue(BTC).avgEntry)
        assertEquals(PaperOrderStatus.FILLED, acct.orders.single().status)
        assertEquals(1_000_000 - 450, acct.cash)
    }

    @Test
    fun limitSell_aboveMarket_rests_thenFillsWhenPriceRises() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("ls", OrderSide.SELL, PaperOrderType.LIMIT, 5, 110), 100)
        assertEquals(PaperOrderStatus.WORKING, acct.orders.single().status)
        acct = PaperEngine.onTick(acct, BTC, 111) // >= 110 triggers
        assertEquals(-5L, acct.positions.getValue(BTC).qty)
        assertEquals(110L, acct.positions.getValue(BTC).avgEntry) // fills at limit, not the tick
    }

    @Test
    fun marketableLimit_fillsImmediatelyAtLimit() {
        val acct = PaperEngine.newAccount(1_000_000)
        // buy limit 120 with market at 100 -> already marketable -> fills at 120
        val after = PaperEngine.placeOrder(acct, order("mb", OrderSide.BUY, PaperOrderType.LIMIT, 2, 120), 100)
        assertEquals(2L, after.positions.getValue(BTC).qty)
        assertEquals(120L, after.positions.getValue(BTC).avgEntry)
        assertEquals(PaperOrderStatus.FILLED, after.orders.single().status)
    }

    @Test
    fun onTick_onlyFillsMatchingSymbol() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("lb", OrderSide.BUY, PaperOrderType.LIMIT, 5, 90, symbolId = BTC), 100)
        // a tick for ETH should not touch the BTC working order
        acct = PaperEngine.onTick(acct, ETH, 50)
        assertEquals(PaperOrderStatus.WORKING, acct.orders.single().status)
        assertTrue(acct.positions.isEmpty())
    }

    // ---- cancel ----

    @Test
    fun cancel_marksWorkingOrderCancelled_leavesNoPosition() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("lb", OrderSide.BUY, PaperOrderType.LIMIT, 5, 90), 100)
        acct = PaperEngine.cancelOrder(acct, "lb")
        assertEquals(PaperOrderStatus.CANCELLED, acct.orders.single().status)
        // a later tick to the limit must NOT fill a cancelled order
        acct = PaperEngine.onTick(acct, BTC, 80)
        assertTrue(acct.positions.isEmpty())
    }

    @Test
    fun cancel_unknownOrFilled_isNoOp() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 1), 100)
        val same = PaperEngine.cancelOrder(acct, "b")     // already filled
        assertEquals(acct, same)
        val same2 = PaperEngine.cancelOrder(acct, "nope") // unknown
        assertEquals(acct, same2)
    }

    // ---- equity / unrealized ----

    @Test
    fun equityAndUnrealized_markLong() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 10), 100)
        // cash 999_000, position 10 @ 100. Mark 150 -> unrealized (150-100)*10 = 500
        val marks = mapOf(BTC to 150L)
        assertEquals(500L, PaperEngine.unrealized(acct, marks))
        // equity = cash + qty*mark = 999_000 + 10*150 = 1_000_500
        assertEquals(1_000_500L, PaperEngine.equity(acct, marks))
    }

    @Test
    fun unrealized_short_profitsAsMarkFalls() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("s", OrderSide.SELL, PaperOrderType.MARKET, 4), 200)
        // short 4 @ 200. Mark 150 -> unrealized (150-200)*(-4) = 200
        assertEquals(200L, PaperEngine.unrealized(acct, mapOf(BTC to 150L)))
    }

    @Test
    fun equity_missingMark_usesCostBasis_noJump() {
        var acct = PaperEngine.newAccount(1_000_000)
        acct = PaperEngine.placeOrder(acct, order("b", OrderSide.BUY, PaperOrderType.MARKET, 10), 100)
        // no mark provided -> position valued at cost basis; equity == startingCash
        assertEquals(1_000_000L, PaperEngine.equity(acct, emptyMap()))
        assertEquals(0L, PaperEngine.unrealized(acct, emptyMap()))
    }

    @Test
    fun zeroQtyOrder_ignored() {
        val acct = PaperEngine.newAccount(1_000_000)
        val after = PaperEngine.placeOrder(acct, order("z", OrderSide.BUY, PaperOrderType.MARKET, 0), 100)
        assertEquals(acct, after)
    }
}
